[ONNX] Passes to reuse existing fake mode if possible

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105769
* __->__ #105764

Fixes #105467, namely the need of setting `aten_graph=True` in `_dynamo.export`
to make fake mode onnx exporter work. 

Previously, `make_fx` called by passes always create new fake mode. Hence it is 
missing out info from `shape_env` recorded during dynamo export. This PR tries
to check and fetch existing fake mode from graph node meta.
Also enable python dispatcher context when calling `make_fx`. This is done in
`_dynamo.export(aten_graph=True)` but was missing in our passes.

BowenBao committed:
[ONNX] Passes to reuse existing 'fake_mode' if possible

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Passes to reuse existing fake mode if possible"


Fixes #105467, namely the need of setting `aten_graph=True` in `_dynamo.export` to make fake mode onnx exporter work. 


[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/_pass.py:
```diff
@@ -163,6 +164,31 @@ def __init__(
         """
         self.module = module
         self.diagnostic_context = diagnostic_context
+        self.fake_mode = self._detect_fake_mode()
```
missing docstring from `Args:` above.

BowenBao replied:
it's not an arg but I'll add it to instance attribute

thiagocrepaldi reviewed (commented):
Does this change do anything other than just remove `aten_graph=` from `torch._dynamo.export` call and add equivalent code to make up for it?

Any new model can be converted or did it fix the FX IR? In that case, should we add a unit test

The resulting code seems to have moved the implementation from the main flag into our code.

BowenBao commented:
Right, there is no extra things. To match aten_graph behavior accurately. 

I think the fundamental reason of such fake mode context setup is to preserve the shape_env info populated during dynamo export.

BowenBao committed:
Update on "[ONNX] Passes to reuse existing fake mode if possible"


Fixes #105467, namely the need of setting `aten_graph=True` in `_dynamo.export`
to make fake mode onnx exporter work. 

Previously, `make_fx` called by passes always create new fake mode. Hence it is 
missing out info from `shape_env` recorded during dynamo export. This PR tries
to check and fetch existing fake mode from graph node meta.
Also enable python dispatcher context when calling `make_fx`. This is done in
`_dynamo.export(aten_graph=True)` but was missing in our passes.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/_pass.py:
```diff
@@ -161,8 +168,33 @@ def __init__(
+            try:
+                fake_tensors.append(node.meta.get("val"))
+            except RuntimeError:
+                continue
```
Comment on when the error is expected?

BowenBao replied:
good catch, I think it was left over when the internal code was calling another function. I'll remove

titaiwangms commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -47,17 +49,27 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
+        if fake_mode is not None:
+            # Using existing fake mode as context, signal `make_fx` that it does not need
+            # to create a new fake mode by passing tracing_mode as "real".
+            tracing_mode = "real"
```
By the way, I think this if/else branch saves us from Fake_mode making dynamic_shapes=False meaningless.

BowenBao replied:
`fake` and `symbolic` both signals `make_fx` to create a new fake mode and use it as context. `real` does nothing.

> By the way, I think this if/else branch saves us from Fake_mode making dynamic_shapes=False meaningless.

Hmm, would it? I think `fake_mode` will not be none for fake export scenario, so it never goes to the branch of selecting `tracing_mode='fake'`.

titaiwangms replied:
I see

titaiwangms commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -102,15 +103,26 @@ def _run(self, *args) -> torch.fx.GraphModule:
+        else:
+            # Existing fake mode not found, signal `make_fx` to create one.
+            fake_mode = contextlib.nullcontext()
+            tracing_mode = "symbolic" if self.enable_dynamic_axes else "fake"
```
It's possible to make this to _pass.py? So we don't write it in passes?

BowenBao replied:
true there is a bit of repetition. I'm slightly leaning towards leaving as is for now. In case there might be more changes to it.

BowenBao commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -102,15 +103,26 @@ def _run(self, *args) -> torch.fx.GraphModule:
-            _allow_fake_constant=self.allow_fake_constant,
-        )(*args)
+
+        fake_mode: Any = self.fake_mode
```
This is for mypy, because later it is overloaded with `contextlib.nullcontext`. I don't want to annotate with ridiculously long type annotations.

thiagocrepaldi replied:
`Any` is hacky because it deceives the type check system Consider `# type: ignore[xxx]` to make it clear that you want to bypass things

BowenBao replied:
Both deceives type check system. I felt this is simple enough code for a single Any. I will change to multiple lines of ignore if folks prefer.

titaiwangms replied:
I don't have strong preference. But even non-annotated better than Any.

BowenBao committed:
Update on "[ONNX] Passes to reuse existing fake mode if possible"


Fixes #105467, namely the need of setting `aten_graph=True` in `_dynamo.export`
to make fake mode onnx exporter work. 

Previously, `make_fx` called by passes always create new fake mode. Hence it is 
missing out info from `shape_env` recorded during dynamo export. This PR tries
to check and fetch existing fake mode from graph node meta.
Also enable python dispatcher context when calling `make_fx`. This is done in
`_dynamo.export(aten_graph=True)` but was missing in our passes.

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -47,17 +49,27 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
+        if fake_mode is not None:
+            # Using existing fake mode as context, signal `make_fx` that it does not need
+            # to create a new fake mode by passing tracing_mode as "real".
+            tracing_mode = "real"
```
This also makes `make_fx` create a graph with static shape.
https://github.com/pytorch/pytorch/issues/105464 has a task to track fake mode + dynamic shape, but a TODO would make finding this integration point easier in the future

BowenBao replied:
This does not make static shape.

This allows reusing the dynamo administrative fake mode, which contains dynamic shape info if dynamic shape is turned on.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -102,15 +103,26 @@ def _run(self, *args) -> torch.fx.GraphModule:
+        if fake_mode is not None:
+            # Using existing fake mode as context, signal `make_fx` that it does not need
+            # to create a new fake mode by passing tracing_mode as "real".
+            tracing_mode = "real"
```
https://github.com/pytorch/pytorch/issues/105464 tracks fake mode + dynamic shape support which might change how `tracing_mode` and `fake_mode` interacts in near future.

I suggest a TODO to help finding this spot in the future when I revisit dyn shape+fake

BowenBao replied:
I can add one if it helps. But be sure to remove the TODO once we get more clarity on fake + dynamic.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -1661,12 +1648,10 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
-            allow_non_fake_inputs=True
-        )
         fake_args = self._fetch_fake_args()
+        fake_mode = self.fake_mode or torch._dynamo.utils.detect_fake_mode(fake_args)
```
it is a bit tricky knowing when we have to use `torch._dynamo.utils.detect_fake_mode` or `torch.onnx._internal.fx._pass._detect_fake_mode`

Maybe we should always use one of them in the onnx exporter to make it uniform? or name our version differently to make clear the distinction, such as `_detect_fake_mode_from_fx_module`

BowenBao replied:
This check with dynamo utils is redundant. I'll remove.

thiagocrepaldi commented on torch/onnx/_internal/fx/_pass.py:
```diff
@@ -161,8 +168,28 @@ def __init__(
+        self.module = module
+        self.fake_mode = self._detect_fake_mode()
+
+    def _detect_fake_mode(self) -> Optional[fake_tensor.FakeTensorMode]:
+        """Detect fake mode from the graph.
+
+        Scan through all nodes in graph and their meta['val'] to detect fake mode.
+        """
+        fake_tensors = [node.meta.get("val") for node in self.module.graph.nodes]
+        return torch._dynamo.utils.detect_fake_mode(fake_tensors)
+
+    def _maybe_fakefy_args(
+        self, fake_mode: Optional[fake_tensor.FakeTensorMode], *args: Any
+    ) -> Tuple[Any, ...]:
+        if fake_mode is None:
+            return args
+        # NB: This should hit the cache if tensors were fakefied before.
+        # E.g., when the fx graph is produced by Dynamo.
+        return tuple(
+            fake_mode.from_tensor(t) if isinstance(t, torch.Tensor) else t for t in args
+        )
```
Maybe move this to a `fake_tensor.py` or `fake_util.py` file? This is a domain specific util not really related to passes only. It is hard to reuse a generic util that is hidden into an unrelated file

This is used by `_pass.py` and maybe (see comment below) should be used by `type_promotion.py`; it might be used at another components when dynamic shape is fully supported

```suggestion
    def _detect_fake_mode(fx_module) -> Optional[fake_tensor.FakeTensorMode]:
        """Detect fake mode from the graph.

        Scan through all nodes in graph and their meta['val'] to detect fake mode.
        """
        fake_tensors = [node.meta.get("val") for node in self.module.graph.nodes]
        return torch._dynamo.utils.detect_fake_mode(fake_tensors)

    def _maybe_fakefy_args(
        fake_mode: Optional[fake_tensor.FakeTensorMode], *args: Any
    ) -> Tuple[Any, ...]:
        if fake_mode is None:
            return args
        # NB: This should hit the cache if tensors were fakefied before.
        # E.g., when the fx graph is produced by Dynamo.
        return tuple(
            fake_mode.from_tensor(t) if isinstance(t, torch.Tensor) else t for t in args
        )
```

BowenBao replied:
Similar to reply at https://github.com/pytorch/pytorch/pull/105764#discussion_r1271612064, but I will do this in a follow up.

BowenBao committed:
address comments on "[ONNX] Passes to reuse existing fake mode if possible"


Fixes #105467, namely the need of setting `aten_graph=True` in `_dynamo.export`
to make fake mode onnx exporter work. 

Previously, `make_fx` called by passes always create new fake mode. Hence it is 
missing out info from `shape_env` recorded during dynamo export. This PR tries
to check and fetch existing fake mode from graph node meta.
Also enable python dispatcher context when calling `make_fx`. This is done in
`_dynamo.export(aten_graph=True)` but was missing in our passes.

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -102,15 +104,29 @@ def _run(self, *args) -> torch.fx.GraphModule:
+            fake_mode = contextlib.nullcontext()  # type: ignore[assignment]
+            tracing_mode = "symbolic" if self.enable_dynamic_axes else "fake"
+
+        assert fake_mode is not None
```
```suggestion
        assert fake_mode is not None, "Cannot detect fake mode."
```

BowenBao replied:
This is impossible for fake_mode to be None, however mypy couldn't figure out.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -47,17 +50,30 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
-            _allow_fake_constant=self.allow_fake_constant,
-        )(*args)
+        assert fake_mode is not None
+        with python_dispatch.enable_python_dispatcher(), fake_mode:
```
out of curiosity, which scenario requires the python dispatcher to be explicitly enabled for the pass (and not for the actual call to `torch._dynamo.export`)?

My understanding it is already enabled by default, so this would be needed if the user explicitly disables it, right? In that scenario, should we re-enable it?

BowenBao replied:
Great question. It is disabled by default. Selectively turned on when one wishes to enable the python dispatcher key in the dispatcher key stack. From export user perspective this should be super low level detail and not exposed.

thiagocrepaldi replied:
I didn't mean visible to the user, but something like

```python

class Exporter:

    def export():
        with python_dispatch.enable_python_dispatcher():
            pre_export_passes()
            torch._dynamo.export()
```

thiagocrepaldi replied:
That is, why we are internally enabling the dispatcher only for a few passes while not to the `fx_tracer.generate_fx()` itself. There is usually a "sync" between `make_fx` and `torch._dynamo.export` calls to keep their fake_modeness identical

BowenBao replied:
Short answer: this matches with aten_graph=True behavior.
Longer answer: `torch._dynamo.export` manages `enable_python_dispatcher` internally. `make_fx` on the other hand can be viewed as a lower level api. While It does have convenience argument `tracing_mode` that sets up fake mode and related configurations inside, one should construct and re-use existing contexts from `torch._dynamo.export` in order to achieve the synchronization regarding fake mode and friends.

BowenBao commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -102,15 +104,29 @@ def _run(self, *args) -> torch.fx.GraphModule:
+            fake_mode = contextlib.nullcontext()  # type: ignore[assignment]
+            tracing_mode = "symbolic" if self.enable_dynamic_axes else "fake"
+
+        assert fake_mode is not None
```
```suggestion
        assert fake_mode is not None  # for mypy
```

BowenBao replied:
Used github commit changes for this. I'm not sure if it gets picked up by ghstack and merge bot, let's find out.

BowenBao replied:
ok that didn't work

BowenBao commented on torch/onnx/_internal/fx/passes/decomp.py:
```diff
@@ -47,17 +50,30 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
-            _allow_non_fake_inputs=True,
-            _allow_fake_constant=self.allow_fake_constant,
-        )(*args)
+        assert fake_mode is not None
```
```suggestion
        assert fake_mode is not None  # for mypy
```

BowenBao committed:
Apply suggestions from code review

BowenBao committed:
Update on "[ONNX] Passes to reuse existing fake mode if possible"


Fixes #105467, namely the need of setting `aten_graph=True` in `_dynamo.export`
to make fake mode onnx exporter work. 

Previously, `make_fx` called by passes always create new fake mode. Hence it is 
missing out info from `shape_env` recorded during dynamo export. This PR tries
to check and fetch existing fake mode from graph node meta.
Also enable python dispatcher context when calling `make_fx`. This is done in
`_dynamo.export(aten_graph=True)` but was missing in our passes.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge
